forecast_model.R

```diff
@@ -3,132 +3,145 @@ library(tidyverse)
 
-#Step 2: Get drivers
 
+# Step 2: Get meterological predictions as drivers
 df_past <- neon4cast::noaa_stage3()
-df_future <- neon4cast::noaa_stage2(cycle = 0)
-sites <- unique(target$site_id)
-
-#Step 3.0: Generate forecasts for each site
-
-forecast <- NULL
-
-#sites <- sites[1:5]
-
- for(i in 1:length(sites)){
-
-  message(paste0("Running site: ", sites[i]))
-
-  # Get site information for elevation
-  site_info <- site_data |> dplyr::filter(field_site_id == sites[i])
 
-  noaa_past <- df_past |>
-    dplyr::filter(site_id == sites[i],
-                  variable == "air_temperature") |>
+## Helper function: for each site, average over predicted 0h horizon ensembles to get 'historic values'
+noaa_mean_historical <- function(df_past, site, var) {
+  df_past |>
+    dplyr::filter(site_id == site,
+                  variable == var) |>
     dplyr::rename(ensemble = parameter) |>
     dplyr::select(datetime, prediction, ensemble) |>
-    dplyr::collect()
-
-  noaa_future <- df_future |>
-    dplyr::filter(site_id == sites[i],
-                  reference_datetime == noaa_date,
-                  datetime >= lubridate::as_datetime(forecast_date),
-                  variable == "air_temperature") |>
-    dplyr::rename(ensemble = parameter) |>
-    dplyr::select(datetime, prediction, ensemble) |>
-    dplyr::collect()
-
-  # Aggregate (to day) and convert units of drivers
-
-  noaa_past_mean <- noaa_past |>
     dplyr::mutate(date = as_date(datetime)) |>
     dplyr::group_by(date) |>
     dplyr::summarize(air_temperature = mean(prediction, na.rm = TRUE),
                      .groups = "drop") |>
     dplyr::rename(datetime = date) |>
-    dplyr::mutate(air_temperature = air_temperature - 273.15)
+    dplyr::mutate(air_temperature = air_temperature - 273.15) |>
+    dplyr::collect()
+}
 
-  noaa_future_site <- noaa_future |>
+## Helper fn: get daily average temperature from each ensemble in future
+noaa_mean_forecast <- function(site, var, reference_date) {
+  endpoint = "data.ecoforecast.org"
+  bucket <- glue::glue("neon4cast-drivers/noaa/gefs-v12/stage1/0/{reference_date}")
+  s3 <- arrow::s3_bucket(bucket, endpoint_override = endpoint, anonymous = TRUE)
+  
+  arrow::open_dataset(s3) |>
+    dplyr::filter(site_id == site,
+                  datetime >= lubridate::as_datetime(forecast_date),
+                  variable == var) |>
+    dplyr::select(datetime, prediction, parameter) |>
     dplyr::mutate(datetime = as_date(datetime)) |>
-    dplyr::group_by(datetime, ensemble) |>
+    dplyr::group_by(datetime, parameter) |>
     dplyr::summarize(air_temperature = mean(prediction), .groups = "drop") |>
     dplyr::mutate(air_temperature = air_temperature - 273.15) |>
-    dplyr::select(datetime, air_temperature, ensemble)
+    dplyr::select(datetime, air_temperature, parameter) |>
+    dplyr::rename(ensemble = parameter) |>
+    dplyr::collect()
+  
+}
+
+# Step 2.5: We'll skip any site that doesn't have both temperature and oxygen
```

Comment:
Why skip sites that don't have both?  Only temperature is needed to build the model and this model can be used to predict temperature only.

Reply:
good question. it just looked to me like that's what you were already doing?

https://github.com/eco4cast/neon4cast-example/blob/5215c30c5a3d71643e4537cf5dae00161c6efdd6/forecast_model.R#L82-L84

I think we'd need to add some extra handling otherwise.  (arguably we should also include chl-a forecasts at the 10 sites that have those measurements?